packages/@uppy/companion-client/src/Provider.ts

```diff
@@ -0,0 +1,445 @@
+    this.provider = opts.provider
+    this.id = this.provider
+    this.name = this.opts.name || getName(this.id)
+    this.pluginId = this.opts.pluginId
+    this.tokenKey = `companion-${this.pluginId}-auth-token`
+    this.companionKeysParams = this.opts.companionKeysParams
+    this.preAuthToken = null
+    this.supportsRefreshToken = opts.supportsRefreshToken ?? true // todo false in next major
+  }
+
+  async headers(): Promise<Record<string, string>> {
+    const [headers, token] = await Promise.all([
+      super.headers(),
+      this.#getAuthToken(),
+    ])
+    const authHeaders: Record<string, string> = {}
+    if (token) {
+      authHeaders['uppy-auth-token'] = token
+    }
+
+    if (this.companionKeysParams) {
+      authHeaders['uppy-credentials-params'] = btoa(
+        JSON.stringify({ params: this.companionKeysParams }),
+      )
+    }
+    return { ...headers, ...authHeaders }
+  }
+
+  onReceiveResponse(response: Response): Response {
+    super.onReceiveResponse(response)
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return response
+    const oldAuthenticated = plugin.getPluginState().authenticated
+    const authenticated = oldAuthenticated
+      ? response.status !== authErrorStatusCode
+      : response.status < 400
+    plugin.setPluginState({ authenticated })
+    return response
+  }
+
+  async setAuthToken(token: string): Promise<any> {
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return
+    // @ts-expect-error don't think we can make core aware that _some_ plugins have a storage
+    plugin.storage.setItem(this.tokenKey, token)
+  }
+
+  async #getAuthToken(): Promise<string | null> {
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return null
+    // @ts-expect-error don't think we can make core aware that _some_ plugins have a storage
+    return plugin.storage.getItem(this.tokenKey)
+  }
+
+  protected async removeAuthToken(): Promise<void> {
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return
+    // @ts-expect-error don't think we can make core aware that _some_ plugins have a storage
+    plugin.storage.removeItem(this.tokenKey)
+  }
+
+  /**
+   * Ensure we have a preauth token if necessary. Attempts to fetch one if we don't,
+   * or rejects if loading one fails.
+   */
+  async ensurePreAuth(): Promise<void> {
+    if (this.companionKeysParams && !this.preAuthToken) {
+      await this.fetchPreAuthToken()
+
+      if (!this.preAuthToken) {
+        throw new Error(
+          'Could not load authentication data required for third-party login. Please try again later.',
+        )
+      }
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars
+  authQuery(data: unknown): Record<string, string> {
+    return {}
+  }
+
+  authUrl({
+    authFormData,
+    query,
+  }: {
+    query: Record<string, string>
+    authFormData: unknown
```

Comment:
Let's put them in the same order they are defined
```suggestion
    authFormData: unknown
    query: Record<string, string>
```